aten/src/ATen/native/sparse/SparseTensorMath.cpp

```diff
@@ -327,27 +332,72 @@ SparseTensor& add_out_sparse_cpu(SparseTensor& r, const SparseTensor& t, const S
         }
     );
 
+    if (r.scalar_type() != commonDtype) {
+      r_values = r_values.to(r.scalar_type());
+    }
+    get_sparse_impl(r)->set_indices_and_values_unsafe(r_indices, r_values);
     get_sparse_impl(r)->set_nnz_and_narrow(r_i);
+
     // TODO: I think it may be possible to track inside the loop and
     // detect when we are uncoalesced (e.g., by observing that an
     // index goes backwards) which may be more precise than using the
     // coalesced flag here.  But this is easy.
-    return r._coalesced_(t_coalesced && s_coalesced);
-  } else {
+    return r._coalesced_(coalesced);
+}
+
+SparseTensor& add_out_sparse_non_contiguous(SparseTensor& r, const SparseTensor& t, const SparseTensor& src, Scalar value, ScalarType commonDtype) {
+    Tensor t_values = t._values().to(commonDtype);
```

Comment:
Why do we need to cast to `commonDtype`? In the end we're not actually performing any addition, so it looks more efficient to:
1) Do the multiplication with the Scalar
2) cast both values to output dtype
3) cat the values together to produce the output values

I think the same suggestion can apply to the CUDA code.

Reply:
I can see that there are a couple of cases (e.g. `half_tensor += double_tensor` or `float_tensor += double_tensor`) where this could be unnecessarily slower as-written. For such examples, your suggestion would be to convert the double_tensor to half instead of the half tensor to double before performing the cat. This would be equivalent (currently) because the final `to()` operation doesn't perform a coalesce, and so wouldn't have any loss of precision _vs_ the current implementation as-written here.

I think a better improvement would be to accept the performance hit but perform the addition at the commonDtype (as we do in dense ops) before converting back to the output type. Having this code be written in a way that explicitly has the operation being performed at the result dtype and not the common dtype seems inconsistent, and like it could cause confusion, even though it could be faster.

i.e., when we call `half_tensor += double_tensor`, the intermediate result at the commonDtype should get coalesced before being converted back to the lower precision result type (but preferably only when the destination dtype is in fact lower precision). This would be similar to what was changed for `div`, where the issue was more pronounced.

I think the best place to fix this though would be in the implementation of sparse `to` (via `copy_sparse_` in SparseTensor.cpp). I could imagine such a change being made eventually anyway and could be beneficial for operations like `float_tensor.copy_(double_tensor)`.

I think what we have here might be good-enough for now though -- I'd prefer to make a change to `to/copy_sparse_` in a follow-up PR in case that raises other issues or testing concerns, and because deciding when to coalesce could be tricky. Thoughts?

Reply:
I don't follow why `half_tensor += double_tensor` is more expensive.

The current code in this PR does:
- cast half_tensor -> double_tensor
- cat double_tensor with double_tensor
- cast double_tensor -> half_tensor.

With my suggestion:
- cast double_tensor -> half_tensor
- cat half_tensor with half_tensor

Reply:
Resolved offline: performance hit seems OK; also, when someone converts a sparse tensor to a lower-precision sparse tensor via .to(...), it might make sense to coalesce or warn that the new tensor might not be usable, but the decision to change those should be motivated by finding use cases for half_tensor += double_tensor.